src/Swarm/Language/Pipeline.hs

```diff
@@ -36,6 +37,11 @@ import Swarm.Language.Typecheck
 import Swarm.Language.Types
 import Witch
 
+-- | A value, or something like a value, along with its
+--   type and requirements
+data Processed val = Processed val Polytype Requirements
```

Comment:
What types will end up being substituted for `val`?  I think I saw `Value` and `Maybe Value`, will there be others?

I wonder if we can come up with a better name than `Processed`.  To me, the name doesn't really help me guess what the definition is.  I guess you named it in parallel with `ProcessedTerm`, but a `ProcessedTerm` actually results from processing a `Term`, whereas that's not the case for a `Processed Value`. (Though perhaps we should rename `ProcessedTerm` as well!)  Also, seeing the type `Processed` made me expect that `ProcessedTerm` is now a type synonym for `Processed Term` or something like that, but it's not.

`Decorated`, perhaps?  `Typeful`?  `Typed`?  I'm open to suggestions.

Reply:
> What types will end up being substituted for val? I think I saw Value and Maybe Value, will there be others?

I'd also not be too surprised if I saw there: `()` or `Term`; `[Var]` (e.g. as a list of suggestions for a typed hole)... It has a functorial feel to it. Obviously I'm not familiar enough with the codebase to have any idea if it's good or not. I also thought maybe there could be a way to have `ProcessedTerm` be part of this already, but I'm not sure.

> I wonder if we can come up with a better name than Processed. 

Completely agree – it's more of a placeholder name, I was hoping for a feedback.

Actually I'm wondering why is it that there's such a hard separation between `Type` and `Requirements`. I guess they "mean" different things, but are they actually ever separated from each other anyway? Doesn't their calculation follow similar paths? Wouldn't they be essentially a decoration over a type in other PLs?

So with that in mind, I do like `Typed`. 

Maybe there could be a name for a generalised type like this, e.g... `AssemblableType`? 

then we could have:

    data AssemblableType = AssemblableType Requirements Polytype
    data Assemblable v = Assemblable AssemblableType v

or even:

    data Assemblable v = AssemblableType Requirements Polytype v
    type AssemblableType = Assemblable ()

And yeah, whatever we decide about the meaning of this should determine which file it goes to, as well.

Reply:
I like `Typed`, that sounds most natural to me. I might consider using a record though:
```Haskell
data Typed v = Typed
  { value :: v,
  , typed :: Polytype
  , requires :: Requirements
  }
```
We use lenses a lot so we might as well use them for this too.


Btw. if you want to get feedback on something in PR, it is a good idea to insert some `TODO` comments so that we don't forget about it. Alternatively, you can make comments on the code here on GitHub or in your editor if you have some plugin for that. :wink:

Reply:
> Actually I'm wondering why is it that there's such a hard separation between Type and Requirements. I guess they "mean" different things, but are they actually ever separated from each other anyway? Doesn't their calculation follow similar paths? Wouldn't they be essentially a decoration over a type in other PLs?

Yes, very good observation.  Right now requirements are calculated separately, after typechecking, but actually there are several bugs related to the fact that it doesn't work quite right (#540, #394) and I want to redesign things so that requirements are part of types, as they should be (#231).

But let's not get too crazy with refactoring in this PR (what you have so far is fine).  I'd rather merge something that works to give us `it`, and then contemplate further refactoring as a separate PR.